pyquil/api/_quantum_computer.py

```diff
@@ -70,6 +73,245 @@ def _get_flipped_protoquil_program(program: Program) -> Program:
+        zero_array = np.zeros((1, num_qubits))
+        one_array = np.ones((1, num_qubits))
+        flip_matrix = np.concatenate((zero_array, one_array), axis=0).astype(int)
+    elif strength == 2:
+        flip_matrix = construct_strength_two_orthogonal_array(num_qubits)
+    elif strength == 3:
+        flip_matrix = construct_strength_three_orthogonal_array(num_qubits)
+
+    return flip_matrix
+
+
+def _next_power_of_2(x):
+    return 1 if x == 0 else 2 ** (x - 1).bit_length()
+
+
+def construct_strength_three_orthogonal_array(num_qubits: int) -> np.ndarray:
+    r"""
+    Given a number of qubits this function returns an Orthogonal Array (OA)
+    on 'n' qubits where n is the next power of two relative to num_qubits.
+
+    Specifically it returns an with the OA(2n, n, 2, 3), where
+
+    OA(N, k, s, t)
+    N: Number of rows, level combinations or runs
+    k: Number of columns, constraints or factors
+    s: Number of symbols or levels
+    t: Strength
+
+    See [OATA] for more details.
+
+    [OATA] Orthogonal Arrays: theory and applications
+           Hedayat, Sloane, Stufken
+           Springer Science & Business Media, 2012.
+           https://dx.doi.org/10.1007/978-1-4612-1478-6
+
+    :param num_qubits: minimum number of qubits the OA should run on.
+    :return: A numpy array representing the OQ with shape N by k
+    """
+    num_qubits_power_of_2 = _next_power_of_2(num_qubits)
+    H = hadamard(num_qubits_power_of_2)
+    Hfold = np.concatenate((H, -H), axis=0)
+    design = ((Hfold + 1) / 2).astype(int)
+    return design
+
+
+def construct_strength_two_orthogonal_array(num_qubits: int) -> np.ndarray:
+    r"""
+    Given a number of qubits this function returns an Orthogonal Array (OA) on 'n-1' qubits
+    where n-1 is the next integer lambda so that 4*lambda -1 is larger than num_qubits.
+
+    Specifically it returns an with the OA(n, n − 1, 2, 2), where
+
+    OA(N, k, s, t)
+    N: Number of rows, level combinations or runs
+    k: Number of columns, constraints or factors
+    s: Number of symbols or levels
+    t: Strength
+
+    See [OATA] for more details.
+
+    [OATA] Orthogonal Arrays: theory and applications
+           Hedayat, Sloane, Stufken
+           Springer Science & Business Media, 2012.
+           https://dx.doi.org/10.1007/978-1-4612-1478-6
+
+    :param num_qubits: minimum number of qubits the OA should run on.
+    :return: A numpy array representing the OQ with shape N by k
+    """
+    # next line will break post denali at 275 qubits
+    # valid_num_qubits = 4 * lambda - 1
+    valid_numbers = [4 * lam - 1 for lam in range(1, 70)]
+    # 4 * lambda
+    four_lam = int(min(x for x in valid_numbers if x >= num_qubits) + 1)
+    H = hadamard(_next_power_of_2(four_lam))
+    # The minus sign in front of H fixes the 0 <-> 1 inversion relative to the reference [OATA]
+    design = ((-H[1:int(four_lam), 0:int(four_lam)] + 1) / 2).astype(int)
```

Comment:
no need for int casting

Reply:
I removed the inner int casting. I am still doing it to the whole array.

- [x] todo